core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java

```diff
@@ -139,9 +166,14 @@ private void reroute() {
             if (lifecycle.stopped()) {
                 return;
             }
+            if (rerouting.compareAndSet(false, true) == false) {
+                logger.trace("already has pending reroute, ignoring");
+                return;
+            }
             clusterService.submitStateUpdateTask(CLUSTER_UPDATE_TASK_SOURCE, Priority.HIGH, new ClusterStateUpdateTask() {
```

Comment:
I will protected against it, good point, I will also add it to the gateway allocator where we do something similar